lnwallet/rpcwallet/sign_coordinator.go

```diff
@@ -0,0 +1,913 @@
+	}
+
+	// We initialize the atomic nextRequestID to the handshakeRequestID, as
+	// requestID 1 is reserved for the initial handshake by the remote
+	// signer.
+	s.nextRequestID.Store(handshakeRequestID)
+
+	return s
+}
+
+// Run starts the SignCoordinator and blocks until the remote signer
+// disconnects, the SignCoordinator is shut down, or an error occurs.
+func (s *SignCoordinator) Run(stream StreamServer) error {
+	s.mu.Lock()
+
+	select {
+	case <-s.quit:
+		s.mu.Unlock()
+		return ErrShuttingDown
+
+	case <-s.doneReceiving:
+		s.mu.Unlock()
+		return ErrNotConnected
+
+	default:
+	}
+
+	s.wg.Add(1)
+	defer s.wg.Done()
+
+	// If we already have a stream, we error out as we can only have one
+	// connection throughout the lifetime of the SignCoordinator.
+	if s.stream != nil {
+		s.mu.Unlock()
+		return ErrMultipleConnections
+	}
+
+	s.stream = stream
+
+	s.mu.Unlock()
+
+	// The handshake must be completed before we can start sending requests
+	// to the remote signer.
+	err := s.handshake(stream)
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Remote signer connected")
+	close(s.clientConnected)
+
+	// Now let's start the main receiving loop, which will receive all
+	// responses to our requests from the remote signer!
+	// We start the receiving loop in a goroutine to ensure that this
+	// function exits if the SignCoordinator is shut down (i.e. the s.quit
+	// channel is closed). Returning from this function will cause the
+	// stream to be closed, which in turn will cause the receiving loop to
+	// exit.
+	s.wg.Add(1)
+	go s.StartReceiving()
+
+	select {
+	case err := <-s.receiveErrChan:
+		return err
+
+	case <-s.quit:
+		return ErrShuttingDown
+
+	case <-s.doneReceiving:
+		return ErrNotConnected
+	}
+}
+
+// Stop shuts down the SignCoordinator and waits until the main receiving loop
+// has exited and all pending requests have been terminated.
+func (s *SignCoordinator) Stop() {
+	log.Infof("Stopping Sign Coordinator")
+	defer log.Debugf("Sign coordinator stopped")
+
+	// We lock the mutex before closing the quit channel to ensure that we
+	// can't get a concurrent request into the SignCoordinator while we're
+	// stopping it. That will ensure that the s.wg.Wait() call below will
+	// always wait for any ongoing requests to finish before we return.
+	s.mu.Lock()
+
+	close(s.quit)
+
+	s.mu.Unlock()
+
+	s.wg.Wait()
+}
+
+// handshake performs the initial handshake with the remote signer. This must
+// be done before any other requests are sent to the remote signer.
+func (s *SignCoordinator) handshake(stream StreamServer) error {
+	var (
+		registerChan     = make(chan *signerResponse)
+		registerDoneChan = make(chan struct{})
+		errChan          = make(chan error)
+	)
+
+	// Create a context with a timeout using the context from the stream as
+	// the parent context. This ensures that we'll exit if either the stream
+	// is closed by the remote signer or if we time out.
+	ctxt, cancel := context.WithTimeout(
```

Comment:
Should we also listen on `s.quit` everywhere? We could use [`fn.ContextGuard`](https://github.com/lightningnetwork/lnd/blob/master/fn/context_guard.go) for this, which has a `WithCtxQuitCustomTimeout` method that would fit well here.

Reply:
Although I guess in this specific case we do listen for `quit` on the outer goroutine, which then cancels the context as well. So probably wouldn't give us much here.

Reply:
> Should we also listen on s.quit everywhere? We could use [fn.ContextGuard](https://github.com/lightningnetwork/lnd/blob/master/fn/context_guard.go) for this, which has a WithCtxQuitCustomTimeout method that would fit well here.

Nice, TIL about the `fn.ContextGuard`! Seems very useful.

> Although I guess in this specific case we do listen for quit on the outer goroutine, which then cancels the context as well. So probably wouldn't give us much here.

Ultimately I didn't change this though due to this comment. I'll update this though if you think it makes the code clearer and more readable.